### PR TITLE
Remove obsolete code

### PR DESCRIPTION
**Pull request by JimBobSquarePants**
### Prerequisites

- [x] I have written a descriptive pull-request title
- [x] I have verified that there are no overlapping [pull-requests](https://github.com/SixLabors/ImageSharp/pulls) open
- [x] I have verified that I am following the existing coding patterns and practice as demonstrated in the repository. These follow strict Stylecop rules :cop:.
- [x] I have provided test coverage for my change (where applicable)

### Description

Removes all obsolete conditions, obsolete methods, and fixes all compiler warnings.

### CODE REVIEW

File: src/ImageSharp/ColorSpaces/Companding/SRgbCompanding.cs [hunk 1]
```diff
@@ -84,19 +82,17 @@ public static class SRgbCompanding
...
-#if SUPPORTS_RUNTIME_INTRINSICS
             if (Avx2.IsSupported && vectors.Length >= 2)
             {
                 CompandAvx2(vectors, ExpandTable);
 
                 if (Numerics.Modulo2(vectors.Length) != 0)
                 {
                     // Vector4 fits neatly in pairs. Any overlap has to be equal to 1.
-                    Expand(ref MemoryMarshal.GetReference(vectors.Slice(vectors.Length - 1)));
+                    Expand(ref MemoryMarshal.GetReference(vectors[^1..]));
```
Comment (gfoidl) [inline src/ImageSharp/ColorSpaces/Companding/SRgbCompanding.cs:92]: The old version may be more performant, unfortunately. This is tracked in https://github.com/dotnet/roslyn/issues/43598, and you can see the difference in [this sharplab](https://sharplab.io/#v2:EYLgtghglgdgPgAQEwEYCwAoBBmABM3AYVwG9NcL88BlABwhgB5gBPAFwFMA+XAWRQAUdBs3bdcAZ3owAlLgC8PKQwDaAPRQA6TQF0A3OUo5cwpq049eSIdNEXJ0uYocNN1ADZQAxhwHKYmgAyHDAA5mwAFrgAtLgoMgYYAL5AA=).

If it's not perf-sensitive, then got with the new code, as it's cleaner to read.
If Roslyn tackles that issue, perf will increase "for free".

File: src/ImageSharp/Formats/Jpeg/Components/Block8x8.cs [hunk 2]
```diff
@@ -271,12 +267,7 @@ public void TransposeInplace()
             // row #6
             Swap(ref Unsafe.Add(ref elemRef, 55), ref Unsafe.Add(ref elemRef, 62));
 
-            static void Swap(ref short a, ref short b)
-            {
-                short tmp = a;
-                a = b;
-                b = tmp;
-            }
+            static void Swap(ref short a, ref short b) => (b, a) = (a, b);
```
Comment (gfoidl) [inline src/ImageSharp/Formats/Jpeg/Components/Block8x8.cs:270]: Just as info:
Resulting machine code will be the same, but the IL-size is with the tuple-based approach a bit larger than with the local (old version).

I don't think it matters here, but potential drawbacks
* JIT may give up inlining if any budget may exceed earlier --> add `MethodImpl.AggressiveInlining` (or the ImageSharp-equivalent) to be on the safe side?
* on products where IL size matters this could have a negative effect (TBH I don't know if two bytes more or less count here, as the reduction overall is much bigger that this PR does, but just to point it out)

File: src/ImageSharp/Formats/Png/PngDecoderCore.cs [hunk 3]
```diff
@@ -1042,7 +1041,7 @@ private void ReadCompressedTextChunk(ImageMetadata baseMetadata, PngMetadata met
             }
 
             int zeroIndex = data.IndexOf((byte)0);
-            if (zeroIndex < PngConstants.MinTextKeywordLength || zeroIndex > PngConstants.MaxTextKeywordLength)
+            if (zeroIndex is < PngConstants.MinTextKeywordLength or > PngConstants.MaxTextKeywordLength)
```
Comment (gfoidl) [inline src/ImageSharp/Formats/Png/PngDecoderCore.cs:1044]: Just FYI: https://github.com/dotnet/roslyn/issues/60534

I'd leave the code as is, as it's basically the same as now just nicer to read. When Roslyn got that addressed the perf-goodness comes "for free".

File: src/ImageSharp/Formats/Webp/Lossy/QuantEnc.cs [hunk 4]
```diff
@@ -77,9 +75,7 @@ public static void PickBestIntra16(Vp8EncIterator it, ref Vp8ModeScore rd, Vp8Se
 
                 if (mode == 0 || rdCur.Score < rdBest.Score)
                 {
-                    Vp8ModeScore tmp = rdCur;
-                    rdCur = rdBest;
-                    rdBest = tmp;
+                    (rdBest, rdCur) = (rdCur, rdBest);
```
Comment (gfoidl) [inline src/ImageSharp/Formats/Webp/Lossy/QuantEnc.cs:78]: Use the swap-method from above (where I commented) instead (needs to be made generic then)?
Or drop the swap-method and write the one-line directly?

I don't have a strong opinion what's better here, but it should be consistent.
Leaning towards swap-method, as if there's a super nice trick once on how to swap more efficiently only one place needs to be updated.

**Review by gfoidl — COMMENTED** [review]
Changes per-se LGTM, but I expect there will be some perf-regression. See comment about Span.Slice vs Range operator.
Maybe you can write a comment on the Roslyn-issue to help them prioritize that one 😉

File: src/ImageSharp/ColorSpaces/Companding/SRgbCompanding.cs (hunk 1 above)
Comment (JimBobSquarePants) [reply src/ImageSharp/ColorSpaces/Companding/SRgbCompanding.cs:92]: 🤦‍♂️ This is so frustrating. Why introduce an analyser that promotes worse behaviour?

File: src/ImageSharp/Formats/Jpeg/Components/Block8x8.cs (hunk 2 above)
Comment (JimBobSquarePants) [reply src/ImageSharp/Formats/Jpeg/Components/Block8x8.cs:270]: Another poor analyser suggestion 😔

File: src/ImageSharp/Formats/Webp/Lossy/QuantEnc.cs (hunk 4 above)
Comment (JimBobSquarePants) [reply src/ImageSharp/Formats/Webp/Lossy/QuantEnc.cs:78]: 👍 good idea

File: src/ImageSharp/Formats/Jpeg/Components/Block8x8.cs (hunk 2 above)
Comment (gfoidl) [reply src/ImageSharp/Formats/Jpeg/Components/Block8x8.cs:270]: I'm no compiler engineer*, but for me this pattern is easy to recognize, so "optimal" code should be emitted, even if Roslyn isn't a optimizing compiler (that's the role of the JIT), but regarding IL-size optimizations count towards Roslyn.

\* so I have somewhat limited understanding on how things work there

File: src/ImageSharp/Formats/Webp/Lossy/Vp8EncIterator.cs [hunk 5]
```diff
@@ -754,7 +755,9 @@ public void MakeChroma8Preds()
 
         public void SwapOut()
         {
+#pragma warning disable IDE0180 // Use tuple to swap values
```
Comment (dlemstra) [inline src/ImageSharp/Formats/Webp/Lossy/Vp8EncIterator.cs:758]: Maybe we could add an extra overload to `RuntimeUtility.Swap`?

File: src/ImageSharp/Common/Helpers/RuntimeUtility.cs [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+        /// Swaps the two references.
+        /// </summary>
+        /// <typeparam name="T">The type to swap.</typeparam>
+        /// <param name="a">The first item.</param>
+        /// <param name="b">The second item.</param>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static void Swap<T>(ref T a, ref T b)
+        {
+            // Tuple swap uses 2 more IL bytes
+#pragma warning disable IDE0180 // Use tuple to swap values
```
Comment (dlemstra) [inline src/ImageSharp/Common/Helpers/RuntimeUtility.cs:24]: Maybe do this on file level instead of the two spots where this is happening?

File: src/ImageSharp/Formats/Webp/Lossy/Vp8EncIterator.cs (hunk 5 above)
Comment (JimBobSquarePants) [reply src/ImageSharp/Formats/Webp/Lossy/Vp8EncIterator.cs:758]: You mean one that takes a `T[]`? The problem I found was that

```c#
RuntimeUtility.Swap(this.YuvOut, this.YuvOut2);
```

Would not work and creating local field references requires a lot more code than we have already.

File: src/ImageSharp/Common/Helpers/RuntimeUtility.cs (hunk 6 above)
Comment (JimBobSquarePants) [reply src/ImageSharp/Common/Helpers/RuntimeUtility.cs]: Good shout

**Review by antonfirsov — COMMENTED** [review]
Looks good, but somewhat concerned about `var` rules.

File: src/ImageSharp/PixelFormats/PixelBlenders/PorterDuffFunctions.Generated.tt [hunk 7]
```diff
@@ -13,10 +13,6 @@
 
 // <auto-generated />
 
-<#
-// Note use of MethodImplOptions.NoInlining. We have tests that are failing on certain architectures when
```
[the remark this answers is not in the document]
Comment (JimBobSquarePants) [reply src/ImageSharp/PixelFormats/PixelBlenders/PorterDuffFunctions.Generated.tt:17]: Yep. All fixed now 😀

File: src/ImageSharp/Common/Helpers/SimdUtils.ExtendedIntrinsics.cs [hunk 8]
```diff
@@ -160,12 +153,10 @@ internal static void ByteToNormalizedFloat(ReadOnlySpan<byte> source, Span<float
                     Vector<uint> w2 = ConvertToUInt32(f2);
                     Vector<uint> w3 = ConvertToUInt32(f3);
 
-                    Vector<ushort> u0 = Vector.Narrow(w0, w1);
-                    Vector<ushort> u1 = Vector.Narrow(w2, w3);
+                    var u0 = Vector.Narrow(w0, w1);
```
[the remark this answers is not in the document]
Comment (JimBobSquarePants) [reply src/ImageSharp/Common/Helpers/SimdUtils.ExtendedIntrinsics.cs:156]: I’ve really struggled to get editorconfig to do what I want here (var rules are actually unchanged). This seems to fall under the “type is apparent” rule. There’s actually been a warning sitting in the code there for years, it just was ignored.

[the remark this answers is not in the document]
Comment (JimBobSquarePants) [reply src/ImageSharp/Common/Helpers/SimdUtils.ExtendedIntrinsics.cs:156]: Why arrays aren’t considered apparent I simply do not know.

[the remark this answers is not in the document]
Comment (antonfirsov) [reply src/ImageSharp/Common/Helpers/SimdUtils.ExtendedIntrinsics.cs:156]: I don't think it's "apparent" when a non-generic type returns a variable with a generic type of the same name. The lack of the type argument hurts readability, especially with API-s like `Vector.Narrow`. Is there a way to fine-tune this rule? What if we disable it entirely?

[the remark this answers is not in the document]
Comment (JimBobSquarePants) [reply src/ImageSharp/Common/Helpers/SimdUtils.ExtendedIntrinsics.cs:156]: I'm gonna create an empty project and see if I can get the rules working using the VS editorconfig editor and a fresh file.

[the remark this answers is not in the document]
Comment (JimBobSquarePants) [reply src/ImageSharp/Common/Helpers/SimdUtils.ExtendedIntrinsics.cs:156]: Here's var turned off everywhere. It's verbose but at least it's consistent.

image

Here's with "when type is apparent"

This one is annoying because the array declaration should be apparent. It's probably the closes to what we want though.

image

Here's "when type is apparent" and "built in types"

This one is annoying because the double and int primitives are not apparent in my opinion since they require the reader to know the syntax declaration rules on the right hand side.

image

Which do you prefer?

[the remark this answers is not in the document]
Comment (antonfirsov) [reply src/ImageSharp/Common/Helpers/SimdUtils.ExtendedIntrinsics.cs:156]: I may be very oldscool, but for me the first one is the lesser evil. I wonder what do others think @SixLabors/core?

This rule/analyzer ([IDE0007?](https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/style-rules/ide0007-ide0008)) really sucks. I wonder if there is a better way to raise awareness about it than posting on https://developercommunity.visualstudio.com/ which is a straight way to oblivion.

[the remark this answers is not in the document]
Comment (JimBobSquarePants) [reply src/ImageSharp/Common/Helpers/SimdUtils.ExtendedIntrinsics.cs:156]: I agree and with the forcing of target type new we handle known types. I'll implement that.

There's actually 2 issues in Roslyn around this. (One closed as a duplicate)

https://github.com/dotnet/roslyn/issues/23714
https://github.com/dotnet/roslyn/issues/55722

I wish I could do this though. It's really annoying that I can't seem to find a way.
```c#
byte[] b = [10];
```

File: src/Directory.Build.props [hunk 9]
```diff
@@ -21,6 +21,10 @@
     <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
   </PropertyGroup>
 
+  <PropertyGroup>
+    <CodeAnalysisRuleSet>..\ImageSharp.ruleset</CodeAnalysisRuleSet>
```
Comment (dlemstra) [inline src/Directory.Build.props:25]: I tend to only use forward slashes now because Windows now also supports that.

File: src/ImageSharp/Formats/Png/PngHeader.cs [hunk 10]
```diff
@@ -116,7 +116,7 @@ public void Validate()
         /// <param name="buffer">The buffer to write to.</param>
         public void WriteTo(Span<byte> buffer)
         {
-            BinaryPrimitives.WriteInt32BigEndian(buffer.Slice(0, 4), this.Width);
+            BinaryPrimitives.WriteInt32BigEndian(buffer[..4], this.Width);
             BinaryPrimitives.WriteInt32BigEndian(buffer.Slice(4, 4), this.Height);
```
Comment (dlemstra) [inline src/ImageSharp/Formats/Png/PngHeader.cs:120]: Did you keep this a `Slice` on purpose?

File: src/ImageSharp/Formats/Webp/Lossy/Vp8Encoding.cs [hunk 11]
```diff
@@ -668,62 +654,62 @@ public static void FTransformWht(Span<short> input, Span<short> output, Span<int
         // luma 16x16 prediction (paragraph 12.3).
         public static void EncPredLuma16(Span<byte> dst, Span<byte> left, Span<byte> top)
         {
-            DcMode(dst.Slice(I16DC16), left, top, 16, 16, 5);
-            VerticalPred(dst.Slice(I16VE16), top, 16);
-            HorizontalPred(dst.Slice(I16HE16), left, 16);
-            TrueMotion(dst.Slice(I16TM16), left, top, 16);
+            DcMode(dst[..], left, top, 16, 16, 5);
```
Comment (dlemstra) [inline src/ImageSharp/Formats/Webp/Lossy/Vp8Encoding.cs:657]: `I16DC16` seems to be missing here?

File: src/ImageSharp/PixelFormats/PixelImplementations/PixelOperations/Rgb24.PixelOperations.cs [hunk 12]
```diff
@@ -34,7 +34,7 @@ internal partial class PixelOperations : PixelOperations<Rgb24>
                 int count = redChannel.Length;
                 GuardPackFromRgbPlanes(greenChannel, blueChannel, destination, count);
 
-                SimdUtils.PackFromRgbPlanes(configuration, redChannel, greenChannel, blueChannel, destination);
+                SimdUtils.PackFromRgbPlanes(redChannel, greenChannel, blueChannel, destination);
```
Comment (dlemstra) [inline src/ImageSharp/PixelFormats/PixelImplementations/PixelOperations/Rgb24.PixelOperations.cs:37]: It looks like we can also remove the `configuration` argument in the next major release?

**Review by dlemstra — COMMENTED** [review]
Made a couple extra commits and I think we are good to go now for a merge.

File: src/ImageSharp/Formats/Webp/Lossy/Vp8Encoding.cs (hunk 11 above)
Comment (JimBobSquarePants) [reply src/ImageSharp/Formats/Webp/Lossy/Vp8Encoding.cs:657]: I'll re read the change.

File: src/ImageSharp/Formats/Png/PngHeader.cs (hunk 10 above)
Comment (JimBobSquarePants) [reply src/ImageSharp/Formats/Png/PngHeader.cs:120]: It wasn't suggested by the IDE. There's situations where is wont be due to performance differences. Maybe this is one of them?

File: src/ImageSharp/Color/Color.cs [hunk 13]
```diff
@@ -286,14 +287,10 @@ public TPixel ToPixel<TPixel>()
...
         /// <typeparam name="TPixel">The pixel type to convert to.</typeparam>
-        /// <param name="configuration">The configuration.</param>
         /// <param name="source">The source color span.</param>
         /// <param name="destination">The destination pixel span.</param>
         [MethodImpl(InliningOptions.ShortMethod)]
-        public static void ToPixel<TPixel>(
-            Configuration configuration,
-            ReadOnlySpan<Color> source,
-            Span<TPixel> destination)
+        public static void ToPixel<TPixel>(ReadOnlySpan<Color> source, Span<TPixel> destination)
```
Comment (dlemstra) [inline src/ImageSharp/Color/Color.cs:293]: This is a breaking change in the Api? Should we not limit this PR to only remove non breaking changes?

File: src/ImageSharp/ColorSpaces/Conversion/Implementation/Converters/HunterLabToCieXyzConverter.cs [hunk 14]
```diff
@@ -9,15 +9,15 @@ namespace SixLabors.ImageSharp.ColorSpaces.Conversion
     /// <summary>
     /// Color converter between <see cref="HunterLab"/> and <see cref="CieXyz"/>
     /// </summary>
-    internal sealed class HunterLabToCieXyzConverter : CieXyzAndHunterLabConverterBase
+    internal class HunterLabToCieXyzConverter : CieXyzAndHunterLabConverterBase
```
Comment (dlemstra) [inline src/ImageSharp/ColorSpaces/Conversion/Implementation/Converters/HunterLabToCieXyzConverter.cs:12]: Why no longer sealed?

File: src/ImageSharp/Color/Color.cs (hunk 13 above)
Comment (antonfirsov) [reply src/ImageSharp/Color/Color.cs:293]: I added `Configuration` to the parameter list to enable the optimization of this method by utilizing the `PixelOperations<TPixel>.FromRgba64()` batch conversion method. The abstract `PixelOperations<TPixel>` methods take `Configuration` for the case the implementation decides to refer to `Configuration.MemoryAllocator` to allocate temporary buffers.

If we don't like this pattern, we can open an issue to re-evaluate the design, after carefully analyzing our current conversion methods, which no longer need special logic for .NET Framework, though we may need to account for non-Intel, non-coreclr targets.

But I would avoid making sneaky public API changes in a big refactor PR. Is this the only one ore is there more?

File: src/ImageSharp/ColorSpaces/Conversion/Implementation/Converters/HunterLabToCieXyzConverter.cs (hunk 14 above)
Comment (JimBobSquarePants) [reply src/ImageSharp/ColorSpaces/Conversion/Implementation/Converters/HunterLabToCieXyzConverter.cs]: Accident. Will reverse

File: src/ImageSharp/Color/Color.cs (hunk 13 above)
Comment (JimBobSquarePants) [reply src/ImageSharp/Color/Color.cs:293]: Let's re-add it for now then and add a TODO in the code so intent is known. 

>Is this the only one ore is there more?
Some of the methods on `ColorSpaceConverter` are now static. I want to keep them like that though.

File: src/ImageSharp/Formats/Webp/Lossy/Vp8Encoding.cs (hunk 11 above)
Comment (JimBobSquarePants) [reply src/ImageSharp/Formats/Webp/Lossy/Vp8Encoding.cs]: No need for a slice at all so just pass `dst`

File: src/ImageSharp/Formats/Jpeg/Components/Decoder/ArithmeticScanDecoder.cs [hunk 15]
```diff
@@ -306,15 +306,16 @@ private ArithmeticStatistics CreateOrGetStatisticsBin(bool dc, int identifier, b
...
-            var statistic = new ArithmeticStatistics(dc, identifier);
+            ArithmeticStatistics statistic = new(dc, identifier);
             this.statistics.Add(statistic);
             return statistic;
         }
 
         private void ParseBaselineData()
         {
-            foreach (ArithmeticDecodingComponent component in this.components)
+            for (int i = 0; i < this.components.Length; i++)
```
Comment (dlemstra) [inline src/ImageSharp/Formats/Jpeg/Components/Decoder/ArithmeticScanDecoder.cs:316]: Why was this changed? Performance?

File: src/ImageSharp/Formats/Jpeg/JpegDecoderCore.cs [hunk 16]
```diff
@@ -751,7 +746,7 @@ private void ProcessApplicationHeaderMarker(BufferedReadStream stream, int remai
             stream.Read(this.temp, 0, JFifMarker.Length);
             remaining -= JFifMarker.Length;
 
-            JFifMarker.TryParse(this.temp, out this.jFif);
+            _ = JFifMarker.TryParse(this.temp, out this.jFif);
```
Comment (dlemstra) [inline src/ImageSharp/Formats/Jpeg/JpegDecoderCore.cs:749]: Should we not handle this?

File: src/ImageSharp/Formats/Png/PngDecoderCore.cs [hunk 17]
```diff
@@ -1590,18 +1561,17 @@ private PngChunkType ReadChunkType()
...
-            {
-                PngThrowHelper.ThrowInvalidChunkType();
 
-                // The IDE cannot detect the throw here.
-                return default;
-            }
+            PngThrowHelper.ThrowInvalidChunkType();
+
+            // The IDE cannot detect the throw here.
+            return default;
```
Comment (dlemstra) [inline src/ImageSharp/Formats/Png/PngDecoderCore.cs:1568]: Do we still need this with the `DoesNotReturn` attribute?

File: src/ImageSharp/Formats/Webp/Lossy/WebpLossyDecoder.cs [hunk 18]
```diff
@@ -252,18 +270,35 @@ private void ParseIntraMode(Vp8Decoder dec, int mbX)
...
                 }
             }
 
             // Hardcoded UVMode decision tree.
-            block.UvMode = (byte)(this.bitReader.GetBit(142) == 0 ? 0 :
-                           this.bitReader.GetBit(114) == 0 ? 2 :
-                           this.bitReader.GetBit(183) != 0 ? 1 : 3);
+            if (this.bitReader.GetBit(142) == 0)
+            {
+                // Hardcoded UVMode decision tree.
```
Comment (dlemstra) [inline src/ImageSharp/Formats/Webp/Lossy/WebpLossyDecoder.cs:281]: Do we really need this comment on every assignment?

File: src/ImageSharp/Metadata/Profiles/Exif/ExifReader.cs [hunk 19]
```diff
@@ -64,11 +66,12 @@ private void GetThumbnail(uint offset)
                 return;
             }
 
-            var values = new List<IExifValue>();
+            List<IExifValue> values = new();
             this.ReadValues(values, offset);
 
-            foreach (ExifValue value in values)
+            for (int i = 0; i < values.Count; i++)
```
Comment (dlemstra) [inline src/ImageSharp/Metadata/Profiles/Exif/ExifReader.cs:72]: Also wondering why this was changed.

File: src/ImageSharp/Metadata/Profiles/ICC/MultiProcessElements/IccMatrixProcessElement.cs [hunk 20]
```diff
@@ -52,13 +52,17 @@ public override bool Equals(IccMultiProcessElement other)
...
         /// <inheritdoc />
         public bool Equals(IccMatrixProcessElement other)
-        {
-            return this.Equals((IccMultiProcessElement)other);
-        }
+            => this.Equals((IccMultiProcessElement)other);
+
+        /// <inheritdoc />
+        public override bool Equals(object obj)
+            => this.Equals(obj as IccMatrixProcessElement);
```
Comment (dlemstra) [inline src/ImageSharp/Metadata/Profiles/ICC/MultiProcessElements/IccMatrixProcessElement.cs:59]: We could also directly do `IccMultiProcessElement` here?

File: src/ImageSharp/Metadata/Profiles/ICC/TagDataEntries/IccXyzTagDataEntry.cs [hunk 21]
```diff
@@ -50,8 +47,13 @@ public override bool Equals(IccTagDataEntry other)
...
         /// <inheritdoc />
         public bool Equals(IccXyzTagDataEntry other)
-        {
-            return this.Equals((IccTagDataEntry)other);
-        }
+            => this.Equals((IccTagDataEntry)other);
+
+        /// <inheritdoc />
+        public override bool Equals(object obj)
+            => this.Equals(obj as IccXyzTagDataEntry);
```
Comment (dlemstra) [inline src/ImageSharp/Metadata/Profiles/ICC/TagDataEntries/IccXyzTagDataEntry.cs:54]: `IccTagDataEntry`?

File: src/ImageSharp/Formats/Png/PngDecoderCore.cs (hunk 17 above)
Comment (JimBobSquarePants) [reply src/ImageSharp/Formats/Png/PngDecoderCore.cs:1568]: It shouldn’t (I don’t think) but I was still getting a warning 🤔

Maybe we need to turn on Nullable Reference Types in a future PR? (Which I would love)

File: src/ImageSharp/Formats/Webp/Lossy/WebpLossyDecoder.cs (hunk 18 above)
Comment (JimBobSquarePants) [reply src/ImageSharp/Formats/Webp/Lossy/WebpLossyDecoder.cs:281]: I think so. I think the reference source is not hard coded and there’s scope for potential enhancement here

File: src/ImageSharp/Metadata/Profiles/Exif/ExifReader.cs (hunk 19 above)
Comment (JimBobSquarePants) [reply src/ImageSharp/Metadata/Profiles/Exif/ExifReader.cs:72]: Because the compiler was complaining that an explicit cast was required for each item. The automatic fix was Linqs cast which was obviously poor from a perf perspective so a for loop was a better fix

File: src/ImageSharp/Metadata/Profiles/ICC/TagDataEntries/IccXyzTagDataEntry.cs (hunk 21 above)
Comment (JimBobSquarePants) [reply src/ImageSharp/Metadata/Profiles/ICC/TagDataEntries/IccXyzTagDataEntry.cs:54]: The cast is the correct type no?

File: src/ImageSharp/Metadata/Profiles/ICC/MultiProcessElements/IccMatrixProcessElement.cs (hunk 20 above)
Comment (JimBobSquarePants) [reply src/ImageSharp/Metadata/Profiles/ICC/MultiProcessElements/IccMatrixProcessElement.cs:59]: The cast is the correct type no?

File: src/ImageSharp/Formats/Png/PngDecoderCore.cs (hunk 17 above)
Comment (dlemstra) [reply src/ImageSharp/Formats/Png/PngDecoderCore.cs:1568]: Would also love to have that enabled. Probably going to be another major PR 😉

File: src/ImageSharp/Metadata/Profiles/ICC/MultiProcessElements/IccMatrixProcessElement.cs (hunk 20 above)
Comment (dlemstra) [reply src/ImageSharp/Metadata/Profiles/ICC/MultiProcessElements/IccMatrixProcessElement.cs:59]: It is correct but the Equals that you are calling does a cast to `IccMultiProcessElement` so you could decide to directly cast instead. Same goes for my other comment.

Comment (JimBobSquarePants) [reply src/ImageSharp/Metadata/Profiles/ICC/MultiProcessElements/IccMatrixProcessElement.cs:59]: Ah right, that depends on an implementation detail of the explicit equality comparer method (which I want to look into. Why are there separate types that do not vary by properties only property values?)

For safety I would always attempt a cast to the concrete type.

File: src/ImageSharp/Formats/Jpeg/JpegDecoderCore.cs (hunk 16 above)
Comment (JimBobSquarePants) [reply src/ImageSharp/Formats/Jpeg/JpegDecoderCore.cs]: Yeah, you're right. Now explicitly throwing `NotSupportedException` since we only support JFIF App0 markers.

**Comment by JimBobSquarePants** [comment]
I'm gonna merge this now. All tests are passing, and I've covered any raised issues.
